// redux-devtools-trace-monitor/lib/react-error-overlay/components/CodeBlock.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var preStyle = {
  position: 'relative',
  display: 'block',
  backgroundColor: '#000',
  padding: '0.5em',
  marginTop: '0.5em',
  marginBottom: '0.5em',
  overflowX: 'auto',
  whiteSpace: 'pre-wrap',
  borderRadius: '0.25rem'
};
var codeStyle = {
  fontFamily: 'Consolas, Menlo, monospace'
};

function CodeBlock(props) {
  var codeBlock = {
    __html: props.codeHTML
  };
  return /*#__PURE__*/_react["default"].createElement("pre", {
    style: preStyle
  }, /*#__PURE__*/_react["default"].createElement("code", {
    style: codeStyle,
    dangerouslySetInnerHTML: codeBlock
  }));
}

var _default = CodeBlock;
exports["default"] = _default;
