// redux-devtools-trace-monitor/lib/react-error-overlay/components/Collapsible.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reduxDevtoolsThemes = require("redux-devtools-themes");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var _collapsibleStyle = {
  color: _reduxDevtoolsThemes.nicinabox.base06,
  backgroundColor: _reduxDevtoolsThemes.nicinabox.base01,
  cursor: 'pointer',
  border: 'none',
  display: 'block',
  width: '100%',
  textAlign: 'left',
  fontSize: '1em',
  padding: '0px 5px',
  lineHeight: '1.5'
};

var collapsibleCollapsedStyle = _objectSpread(_objectSpread({}, _collapsibleStyle), {}, {
  marginBottom: '1.5em'
});

var collapsibleExpandedStyle = _objectSpread(_objectSpread({}, _collapsibleStyle), {}, {
  marginBottom: '0.6em'
});

var Collapsible = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Collapsible, _Component);

  var _super = _createSuper(Collapsible);

  function Collapsible() {
    var _this;

    (0, _classCallCheck2["default"])(this, Collapsible);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      collapsed: undefined
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "toggleCollapsed", function () {
      _this.setState(function (state) {
        return {
          collapsed: !_this.isCollapsed(state)
        };
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "isCollapsed", function (state) {
      return state.collapsed === undefined ? _this.props.collapsedByDefault : state.collapsed;
    });
    return _this;
  }

  (0, _createClass2["default"])(Collapsible, [{
    key: "render",
    value: function render() {
      var count = this.props.children.length;
      var collapsed = this.isCollapsed(this.state);
      return /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement("button", {
        onClick: this.toggleCollapsed,
        style: collapsed ? collapsibleCollapsedStyle : collapsibleExpandedStyle
      }, (collapsed ? '▶' : '▼') + " ".concat(count, " stack frames were ") + (collapsed ? 'collapsed.' : 'expanded.')), /*#__PURE__*/_react["default"].createElement("div", {
        style: {
          display: collapsed ? 'none' : 'block'
        }
      }, this.props.children, /*#__PURE__*/_react["default"].createElement("button", {
        onClick: this.toggleCollapsed,
        style: collapsibleExpandedStyle
      }, "\u25B2 ".concat(count, " stack frames were expanded."))));
    }
  }]);
  return Collapsible;
}(_react.Component);

var _default = Collapsible;
exports["default"] = _default;
