// remotedev-utils/src/index.js
import getParams from 'get-params';

function flatTree(obj, namespace = '') {
  let functions = [];
  Object.keys(obj).forEach(key => {
    const prop = obj[key];
    if (typeof prop === 'function') {
      functions.push({
        name: namespace + (key || prop.name || 'anonymous'),
        func: prop,
        args: getParams(prop),
      });
    } else if (typeof prop === 'object') {
      functions = functions.concat(flatTree(prop, namespace + key + '.'));
    }
  });
  return functions;
}

export function getActionsArray(actionCreators) {
  if (Array.isArray(actionCreators)) return actionCreators;
  return flatTree(actionCreators);
}

/* eslint-disable no-new-func */
export function evalAction(action, actionCreators) {
  if (typeof action === 'string') {
    return (new Function('return ' + action))();
  }

  const actionCreator = actionCreators[action.selected].func;
  const interpretArg = (arg) => (new Function('return ' + arg))();
  let argsToInject = action.args.map(interpretArg);
  if (action.rest) {
    const rest = interpretArg(action.rest);
    if (Array.isArray(rest)) argsToInject = argsToInject.concat(...rest);
    else throw new Error('rest must be an array');
  }
  return actionCreator(...argsToInject);
}
/* eslint-enable */
